compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -801,15 +804,59 @@ static LogicalResult setMatmulPeelingRootConfig(
   SmallVector<int64_t> vectorReductionTileSizes(numTilingDims, 0);
   std::swap(vectorParallelTileSizes.back(), vectorReductionTileSizes.back());
 
+  // The backend struggles to legalize non-power-of-two scalable vectors,
+  // hence the extra rounding up.
+  for (const auto &[index, size] : llvm::enumerate(vectorParallelTileSizes)) {
+    if (!size)
+      continue;
+    vectorParallelTileSizes[index] =
+        roundUpToPow2(size,
+                      /*predicate=*/inputVecScalableTileFlags[index]);
+  }
+
   TileSizesListType tileSizes = {
       SmallVector<int64_t>(distTileSizes), cacheParallelTileSizes,
       cacheReductionTileSizes, vectorParallelTileSizes,
       vectorReductionTileSizes};
   // No need for tiling inner parallel dims.
   tileSizes.emplace_back(numTilingDims, 0);
 
+  // 2. Set scalable flags for all the tiling levels.
+  SmallVector<bool> parallelScalableFlags(inputVecScalableTileFlags.begin(),
+                                          inputVecScalableTileFlags.end());
+  SmallVector<bool> reductionScalableFlags(numTilingDims, false);
+
+  // Ensure there's no zero scalable dims.
+  for (unsigned i = 0; i < numTilingDims; i++) {
+    if (vectorReductionTileSizes[i] == 0)
+      reductionScalableFlags[i] = false;
```

Comment:
Do we need this? It looks like reductionScalableFlags[*] is always false.

We are assuming it is a `[B], M, N, K` kind of matmul, (which is actually not good, but probably fine for now). How about following the logic like other tile sizes? E.g.,

1. we copy `inputVecScalableTileFlags` to `parallelScalableFlags`.
2. initialize `reductionScalableFlags` to all false.
3. swap the last element for `parallelScalableFlags` and `reductionScalableFlags`.

Does it work (and look better)?

```cpp
  SmallVector<bool> parallelScalableFlags(inputVecScalableTileFlags.begin(),
                                          inputVecScalableTileFlags.end());
  SmallVector<bool> reductionScalableFlags(numTilingDims, false);
  std::swap(parallelScalableFlags.back(), reductionScalableFlags.back());
```

Reply:
Yup, that's much nicer, thanks! Sending update version shortly.